Performance and minification improvements

niemyjski opened the pull request.

niemyjski commented on src/controllers/controller.bar.js:
```diff
@@ -299,18 +304,23 @@ module.exports = function(Chart) {
-			rectangle._model.backgroundColor = rectangle.custom && rectangle.custom.backgroundColor ? rectangle.custom.backgroundColor : helpers.getValueAtIndexOrDefault(this.getDataset().backgroundColor, index, this.chart.options.elements.rectangle.backgroundColor);
-			rectangle._model.borderColor = rectangle.custom && rectangle.custom.borderColor ? rectangle.custom.borderColor : helpers.getValueAtIndexOrDefault(this.getDataset().borderColor, index, this.chart.options.elements.rectangle.borderColor);
-			rectangle._model.borderWidth = rectangle.custom && rectangle.custom.borderWidth ? rectangle.custom.borderWidth : helpers.getValueAtIndexOrDefault(this.getDataset().borderWidth, index, this.chart.options.elements.rectangle.borderWidth);
+			model.backgroundColor = custom && custom.backgroundColor ? custom.backgroundColor : helpers.getValueAtIndexOrDefault(dataset.backgroundColor, index, rectangleElementOptions.backgroundColor);
```
There might be a bug here. `dataset` variable was defined but wasn't being used before and I updated it to use it..

simonbrunel commented:
Thanks @niemyjski. FYI, I already reviewed core.scale.js and element.line.js for performance and minification, I'm just waiting for [this PR](https://github.com/chartjs/chartjs-color/pull/1) to be merged. I guess I will have to deal with many merge conflicts :)

simonbrunel commented on .editorconfig:
```diff
@@ -0,0 +1,10 @@
+end_of_line = lf
+charset = utf-8
+trim_trailing_whitespace = true
+insert_final_newline = false
```
Should be `insert_final_newline = true`, right?

niemyjski replied:
I was looking at the code and we didn't have any new lines on all the docs I checked

simonbrunel commented on src/charts/Chart.Radar.js:
```diff
@@ -1,15 +1,9 @@
-
-	var helpers = Chart.helpers;
-
-	var defaultConfig = {
```
`Chart.Scatter` also define [defaultConfig](https://github.com/chartjs/Chart.js/blob/master/src/charts/Chart.Scatter.js#L5), do we really want to inline `defaultConfig` here?

niemyjski replied:
it won't be exposed, it's local to this function.

niemyjski replied:
yeah, it's local to this function and if it's not being used more than once or severely hurting readability it should be inlined.

simonbrunel replied:
You absolutely right, I thought it was a member.

simonbrunel commented on src/controllers/controller.bar.js:
```diff
@@ -112,15 +116,15 @@ module.exports = function(Chart) {
-					borderSkipped: rectangle.custom && rectangle.custom.borderSkipped ? rectangle.custom.borderSkipped : this.chart.options.elements.rectangle.borderSkipped,
-					borderColor: rectangle.custom && rectangle.custom.borderColor ? rectangle.custom.borderColor : helpers.getValueAtIndexOrDefault(this.getDataset().borderColor, index, this.chart.options.elements.rectangle.borderColor),
-					borderWidth: rectangle.custom && rectangle.custom.borderWidth ? rectangle.custom.borderWidth : helpers.getValueAtIndexOrDefault(this.getDataset().borderWidth, index, this.chart.options.elements.rectangle.borderWidth)
+					backgroundColor: custom && custom.backgroundColor ? custom.backgroundColor : helpers.getValueAtIndexOrDefault(dataset.backgroundColor, index, rectangleElementOptions.backgroundColor),
```
For all these kind of check, I would rather do:
```javascript
    var custom = rectangle.custom || {};
    ...
    backgroundColor: custom.backgroundColor? custom.backgroundColor : helpers.getValueAtIndexOrDefault(dataset.backgroundColor, index, rectangleElementOptions.backgroundColor);
```
or
```javascript
    backgroundColor: custom.backgroundColor || helpers.getValueAtIndexOrDefault(dataset.backgroundColor, index, rectangleElementOptions.backgroundColor);
```

What do you think?

niemyjski replied:
updated

simonbrunel commented on src/core/core.controller.js:
```diff
@@ -322,10 +323,11 @@ module.exports = function(Chart) {
-			this.chart.ctx.beginPath();
-			this.chart.ctx.rect(this.chartArea.left, this.chartArea.top, this.chartArea.right - this.chartArea.left, this.chartArea.bottom - this.chartArea.top);
-			this.chart.ctx.clip();
+			var context = this.chart.ctx;
```
I would use `ctx` instead of `context` because `ctx` is used in many other places in the code (and is more compact).

niemyjski replied:
@etimberg @simonbrunel the reason I renamed it context in most places is because it's easier to read and will be minified. Also there were some places where you have this.ctx and naming it ctx would of caused some confusion and broken refactoring. If you wanted to change the ctx local variable it would change all references as it thinks `ctx` and `this.ctx` are the same. Also I made it context all over to be consistent. If you don't like this change I can change it but I already hit this issue with this.ticks and ticks in this code base.

etimberg replied:
@niemyjski That's a fine reason. we can always come back later and make it ctx or something if we really wanted

simonbrunel replied:
I guess the *easier to read* is a personal thing :) (to me it's easier to read many `ctx` than `context`). That's also because I like short names and `ctx` is commonly used. Not sure about the renaming argument, it would mean that we have to always pick local variable names different from class members (e.g `var gridLines = options.gridLines;`, `var scaleLabel = options.scaleLabel;`, etc...)?

Nevermind, if @etimberg is fine with that, I'm as well :)

niemyjski replied:
It's only when you have ctx and this.ctx or ticks and this.ticks that refactoring really fails. They could be pointing to the same thing or an options ticks. Since we are not using typescript which has way better tooling, it fails pretty quick

etimberg replied:
@simonbrunel I'm okay with it since the minifier will get it. In terms of making everything consistent (always using ctx) then I think that can be done later. We'll probably be back in these files soon to do more minification

simonbrunel commented on src/core/core.scale.js:
```diff
@@ -3,6 +3,7 @@
 module.exports = function(Chart) {
 
 	var helpers = Chart.helpers;
+	var globalDefaults = Chart.defaults.global;
```
This doesn't seem to be used?

niemyjski replied:
done

etimberg commented:
Looks pretty good. I think @simonbrunel did a good job looking through as well

niemyjski commented:
I just committed all the changes you guys wanted. The only thing I think needs to be taking a look at is `removeHoverStyle` functions. Please do a search for them and look at the two places I changed in the changes above. both had a dataset variable that wasn't being used and then the code following that declaration was calling `this.getDataset()`. I updated it to use the existing variable and all the tests pass.

etimberg commented:
If we're all agreed, lets merge this :)

niemyjski commented:
Looks good to me :)

simonbrunel commented:
Good work @niemyjski!
